File: scala/120.scala
object Solution {
    def minimumTotal(triangle: List[List[Int]]): Int = {
        val arr = triangle.map(l => l.toArray).toArray
        for (i <- 1 until arr.length) {
            for (j <- 0 until arr(i).length) {
                if (j == 0) arr(i)(j) += arr(i-1)(0)
                else if (j==arr(i).length-1) arr(i)(j) += arr(i-1)(j-1)
                else arr(i)(j) += arr(i-1)(j).min(arr(i-1)(j-1))
            }
        }
        arr.last.min
    }
}

File: leetcodeJava/120.java
package leetcodeJava;

class Solution {
    public int minimumTotal(List<List<Integer>> triangle) {
        var dp = new int[triangle.size()][2];
        var last = 0;
        for (var i = 0; i < triangle.size(); i++) {
            for (var j = 0; j <= i; j++) {
                if (i == 0) dp[j][1-last] = triangle.get(i).get(j);
                else {
                    if (j == 0) {
                        dp[j][1-last] = dp[j][last] + triangle.get(i).get(j);
                    } else if (j==i) {
                        dp[j][1-last] = dp[j-1][last] + triangle.get(i).get(j);
                    } else {
                        dp[j][1-last] = Math.min(dp[j-1][last], dp[j][last]) + triangle.get(i).get(j);
                    }
                }
            }
            last = 1-last;
        }
        var res = 10001;
        for (var i = 0; i < dp.length; i++) {
            if (dp[i][last] < res) res = dp[i][last];
        }
        return res;
    }
}
